### ISSUE TITLE
[Job] Submitting jobs without delay can cause jobs PENDING forever

### ISSUE THREAD

**Issue by Michaelvll** [skypilot-org/skypilot#1125]
The target user tried to continuously submit spot jobs without delay, and after about 100 jobs finished running, there are about 30 jobs keep in PENDING status and no jobs are running. 
![image](https://user-images.githubusercontent.com/6753189/186725960-00fc396e-6d50-40c4-8445-1fd53b6309dc.png)

It seems caused by ray not correctly scheduling the placement groups (0.5 CPU) for the jobs. Here is the logs for `ray status`
```
======== Autoscaler status: 2022-08-25 17:06:44.549228 ========
Node status
---------------------------------------------------------------
Healthy:
 1 ray_head_default
Pending:
 (no pending nodes)
Recent failures:
 (no failures)

Resources
---------------------------------------------------------------
Usage:
 0.0/8.0 CPU (0.0 used of 8.0 reserved in placement groups)
 0.00/30.052 GiB memory
 0.00/15.026 GiB object_store_memory

Demands:
 {'node:10.128.15.204': 0.001}: 78+ pending tasks/actors
 {'CPU': 0.5} * 1 (STRICT_SPREAD): 33+ pending placement groups
```

In the `raylet.err`, there are a lot of following errors:
```
Traceback (most recent call last):
  File "/opt/conda/lib/python3.7/site-packages/ray/workers/default_worker.py", line 216, in <module>
    ray_debugger_external=args.ray_debugger_external,
  File "/opt/conda/lib/python3.7/site-packages/ray/worker.py", line 1624, in connect
    worker.import_thread.start()
  File "/opt/conda/lib/python3.7/site-packages/ray/_private/import_thread.py", line 56, in start
    self.t.start()
  File "/opt/conda/lib/python3.7/threading.py", line 852, in start
    _start_new_thread(self._bootstrap, ())
RuntimeError: can't start new thread
Error in atexit._run_exitfuncs:
Traceback (most recent call last):
  File "/opt/conda/lib/python3.7/site-packages/ray/_private/client_mode_hook.py", line 105, in wrapper
    return func(*args, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/ray/worker.py", line 1160, in shutdown
    disconnect(_exiting_interpreter)
  File "/opt/conda/lib/python3.7/site-packages/ray/worker.py", line 1710, in disconnect
    worker.import_thread.join_import_thread()
  File "/opt/conda/lib/python3.7/site-packages/ray/_private/import_thread.py", line 61, in join_import_thread
    self.t.join()
  File "/opt/conda/lib/python3.7/threading.py", line 1039, in join
    raise RuntimeError("cannot join thread before it is started")
RuntimeError: cannot join thread before it is started
terminate called without an active exception
```

The number of the JobSupervisor matches the number of RUNNING jobs in `ray job list`, which means the JobSupervisor is not leaked.
```
ps aux | grep JobSupervisor| wc -l
127
```

When I tried to connect to the ray cluster with `ray.init('auto')`, the following error occurs.
```
>>> import ray
ray>>> ray.init('auto')
terminate called after throwing an instance of 'std::system_error'
  what():  Resource temporarily unavailable
*** SIGABRT received at time=1661448234 on cpu 5 ***
PC: @     0x7f5fb84ba7bb  (unknown)  raise
    @     0x7f5fb87ec730  (unknown)  (unknown)
    @ 0x61726f706d657420  (unknown)  (unknown)
[2022-08-25 17:23:54,575 E 2559 2559] logging.cc:325: *** SIGABRT received at time=1661448234 on cpu 5 ***
[2022-08-25 17:23:54,575 E 2559 2559] logging.cc:325: PC: @     0x7f5fb84ba7bb  (unknown)  raise
[2022-08-25 17:23:54,576 E 2559 2559] logging.cc:325:     @     0x7f5fb87ec730  (unknown)  (unknown)
[2022-08-25 17:23:54,577 E 2559 2559] logging.cc:325:     @ 0x61726f706d657420  (unknown)  (unknown)
Fatal Python error: Aborted

Stack (most recent call first):
  File "/opt/conda/lib/python3.7/site-packages/ray/worker.py", line 1606 in connect
  File "/opt/conda/lib/python3.7/site-packages/ray/worker.py", line 1109 in init
  File "/opt/conda/lib/python3.7/site-packages/ray/_private/client_mode_hook.py", line 105 in wrapper
  File "<stdin>", line 1 in <module>
Aborted
```

The `dmesg` has the the following message:
```
[13650.878632] hrtimer: interrupt took 21375 ns
[19785.296747] cgroup: fork rejected by pids controller in /user.slice/user-801785486.slice/session-1.scope
[19791.707331] cgroup: fork rejected by pids controller in /user.slice/user-801785486.slice/session-33.scope
[19836.892408] cgroup: fork rejected by pids controller in /user.slice/user-801785486.slice/session-566.scope
[19922.013667] cgroup: fork rejected by pids controller in /user.slice/user-801785486.slice/session-568.scope
[20009.737091] cgroup: fork rejected by pids controller in /user.slice/user-801785486.slice/session-571.scope
[20092.404312] cgroup: fork rejected by pids controller in /user.slice/user-801785486.slice/session-573.scope
[20176.906418] cgroup: fork rejected by pids controller in /user.slice/user-801785486.slice/session-575.scope
[20431.471705] cgroup: fork rejected by pids controller in /user.slice/user-801785486.slice/session-583.scope
[20515.707272] cgroup: fork rejected by pids controller in /user.slice/user-801785486.slice/session-585.scope
[20600.301047] cgroup: fork rejected by pids controller in /user.slice/user-801785486.slice/session-588.scope
[20694.727789] cgroup: fork rejected by pids controller in /user.slice/user-801785486.slice/session-590.scope
[21070.019273] cgroup: fork rejected by pids controller in /user.slice/user-801785486.slice/session-601.scope
[21155.494996] cgroup: fork rejected by pids controller in /user.slice/user-801785486.slice/session-604.scope
[21239.563674] cgroup: fork rejected by pids controller in /user.slice/user-801785486.slice/session-606.scope
[21410.820676] cgroup: fork rejected by pids controller in /user.slice/user-801785486.slice/session-611.scope
[21663.961697] cgroup: fork rejected by pids controller in /user.slice/user-801785486.slice/session-618.scope
[26374.991020] cgroup: fork rejected by pids controller in /user.slice/user-801785486.slice/session-753.scope
[31979.026179] cgroup: fork rejected by pids controller in /user.slice/user-801785486.slice/session-837.scope
```

By checking the task limit of the system we can find that the limitation of the tasks for the current user is `10813`, and the total number of threads created by the user is `10722`, which is very close to the limit. It means that the problem is very likely to be caused by that too many jobs are pending and each job may have 50 threads or so.
```
$ sudo systemd-analyze dump | sed -n "/-> Unit user-801785486.slice:/,/-> Unit /p"| grep -e "TasksMax="
	TasksMax=10813
$ ps -fLu qzeng_salk_edu | wc -l
10722
```

jobs.db and spot_jobs.db: https://gist.github.com/Michaelvll/fa1222c1ccf556c1005430d4d22414c0

**Comment by Michaelvll** [comment]
The problems can also be reproduced by the code in #1127

**Comment by Michaelvll** [comment]
We can probably increase the max task limit for the whole system as mentioned by modifying the system file. https://www.suse.com/support/kb/doc/?id=000015901

**Comment by Michaelvll** [comment]
A weird thing happens. When running the reproduce code from #1127, when the cpunode is firstly launched, it will stuck at job 197 for connecting to the dashboard. However, if we restart the ray cluster, it won't stuck anymore.

The error from the raylet from the first time is that:
```
what(): epoll: Too many open files
```
However, we already enlarged the nofile limits.

**Comment by Michaelvll** [comment]
Another possible reproducible code:

```
sky launch -c repr —cloud gcp ‘’

for i in {1..100}; do
  sky exec -d repr “echo start; sleep 1000000000000000”
done

```

**Comment by Michaelvll** [comment]
After the fixes from #1128 and #1130, the problem should have been solved. Closing this issue for now. Please feel free to re-open this issue, when the problem occurs again @lhqing @rachelzeng98. ; )
